### DISCUSSION TITLE
Apache2 Sample Setup for Reverse Proxy Installation

### DISCUSSION THREAD

**Discussion by rg-ac** [keycloak/keycloak#11640] [Q&A]
### Description

I wasn't able to find a working Apache2 setup. What ever I tried there was still a browser mixed content failure in the account page. May be you can add a sample configuration for Apache2. With NGinx I was able to get it running. The last failed Apache2 setup (Debian 11.1) was:
[..]
```
<IfModule mod_ssl.c>
        <VirtualHost _default_:443>
                DocumentRoot /var/www/html
                Header setIfEmpty X-Forwarded-Proto "https"
                ProxyPreserveHost On

                # ProxyPass adds X-Forwarded-For, X-Forwarded-Host and X-Forwarded-Server
                ProxyPass / http://localhost:8080/
                ProxyPassReverse / http://localhost:8080/
```
[..]

### Motivation

Failed installation/setup.

### Details

Apache2 VirtualHost Sample Setup

**Comment by ZPascal** [comment]
I'm using the following working configuration of the apach2 reverse proxy:

```xml
<IfModule mod_ssl.c>
<VirtualHost *:443>
        ServerName sso.test.com
        ServerAdmin root@localhost
        DocumentRoot /var/www/html

        # don't lose time with IP address lookups
        HostnameLookups Off
        # needed for named virtual hosts
        UseCanonicalName Off

        ProxyPreserveHost On
        RequestHeader set X-Forwarded-Proto "https"
        RequestHeader set X-Forwarded-Port "443"
        ProxyPass / http://localhost:8080/
        ProxyPassReverse / http://localhost:8080/

        ProxyRequests Off

        SecRuleEngine On
        SecRule ARGS:triggerparam "@contains trigger" "id:1234,deny,status:403,msg:'Our trigger rule has triggered'"

        <LocationMatch "(/auth/admin/|/auth/realms/master/)$">
                deny from all
                allow from x.x.x.x

                AuthType Basic
                AuthName "SSO Admin protection"
                AuthUserFile /path/sso/.htpasswd
                require valid-user
        </LocationMatch>

        <LocationMatch "/auth/admin">
                SecRuleRemoveById 911100
                SecRuleRemoveById 949110
                SecRuleRemoveById 980130
        </LocationMatch>

        ErrorLog /var/log/apache2/keycloaktestcom.log
        LogLevel warn
        CustomLog /var/log/apache2/keycloaktestcom.log combined
        ServerSignature Off

SSLCertificateFile /etc/letsencrypt/live/sso.test.com/fullchain.pem
SSLCertificateKeyFile /etc/letsencrypt/live/sso.test.com/privkey.pem
Include /etc/letsencrypt/options-ssl-apache.conf
</VirtualHost>
</IfModule>
```
As you can already see, the admin interface is particularly secured (ModSecurity -> SecRuleEngine & Basic auth), you could possibly ignore that in your scenario. Furthermore, my Keycloak instance is running on version 18.0.0 and within the configuration I set the `http-relative-path=/auth` parameter to be able to restore the old WildFly configuration and not have to migrate my clients.

**Reply by ajquick** [reply]
I have spent a long time writing my VirtualHost proxy redirects. I'm really glad to have found your example as it allowed me to expand on mine. This is mine. It may be a bit more updated for Apache 2.4+ as they changed a few things. I have this working semi-successfully on Cpanel / Cloudlinux with Mod_Security.

````
<VirtualHost *:80>
    ServerName keycloak.domain.com
    ServerAlias www.keycloak.domain.com
    Redirect Permanent / https://keycloak.domain.com/
    RewriteEngine on
    RewriteCond %{SERVER_NAME} =keycloak.domain.com
    RewriteRule ^ https://%{SERVER_NAME}%{REQUEST_URI} [END,NE,R=permanent]
</VirtualHost>

<VirtualHost *:443>
    ServerName keycloak.domain.com
    ServerAlias www.keycloak.domain.com

    HostnameLookups Off
    UseCanonicalName Off

    <Proxy *>
        Require all granted
    </Proxy>

    SSLEngine on
    SSLProxyEngine On
    SSLCertificateFile /var/cpanel/ssl/domain_tls/keycloak.domain.com/certificates
    SSLCertificateKeyFile /var/cpanel/ssl/domain_tls/keycloak.domain.com/combined
 
    RemoteIPHeader X-Forwarded-For
    RequestHeader set "X-forwarded-proto" "https"
    RequestHeader set X-Forwarded-Port "443"

    ProxyPreserveHost On

    ProxyPass / http://127.0.0.1:8080/
    ProxyPassReverse / http://127.0.0.1:8080/

    ProxyRequests Off

    <Location />
        Require all denied   
    </Location>

    <LocationMatch "(/$|/admin|/welcome|/metrics|/health)">
        Require all denied        
        Require ip 0.0.0.0
        Require ip MY_IP
        Require ip 127.0.0.0/8
        Require ip 192.168.0.0/16
        Require ip SERVER_IP
    </LocationMatch>

    <LocationMatch "(/js|/realms|/resources/|/robots.txt)">
        Require all granted 
    </LocationMatch>

    SecRuleEngine On

    <LocationMatch "/admin">
        SecRuleRemoveById 931100
        SecRuleRemoveById 949110
        SecRuleRemoveById 980130
    </LocationMatch>

</VirtualHost>
````

I can't access anything on https://127.0.0.1:8443, so I have to use http://127.0.0.1:8080... the same as you. I'm using proxy mode set to Edge. Haven't been able to get passthrough or reencrypt to work.

Any ideas? Or should I not bother since it works?

Error when trying to use https and 8443:

    Proxy Error
    The proxy server received an invalid response from an upstream server.
    The proxy server could not handle the request
    
    Reason: Error reading from remote server

**Reply by ZPascal** [reply]
Which Keycloak version are you using? I think there was a bug in a version before v18.0.0. Do you have a requirement to use TLS between the proxy and the Keycloak instance?

**Reply by darius-m** [reply]
It's not quite related to the topic, but you have an issue with allowed IPs in the example above - the `192.0.0.0/8` are public IPs, with the exception of `192.168.0.0/16`.

**Reply by ajquick** [reply]
Thanks for the correction @darius-m!

> Which Keycloak version are you using? I think there was a bug in a version before v18.0.0. Do you have a requirement to use TLS between the proxy and the Keycloak instance?

@ZPascal I'm using 18.0.1. This is my first time using Keycloak and it has quite the learning curve. My settings are as follows:

````
db=mariadb
db-username=USER
db-password=PASSWORD
proxy=edge
#hostname-admin=127.0.0.1
hostname=keycloak.domain.com
hostname-strict=false
hostname-strict-https=false
http-host=127.0.0.1
https-certificate-file=/var/cpanel/ssl/domain_tls/keycloak.domain.com/certificates
https-certificate-key-file=/var/cpanel/ssl/domain_tls/keycloak.domain.com/combined
https-client-auth=required
https-protocols=TLSv1.3,TLSv1.2
https-port=8443
http-enabled=true
log-level=info,warn,error
log=file
log-file=/home/keycloak/logs/keycloak.log
````

Basically. I have no idea. It has been really hard to understand what is a setting from versions before 17 and versions after 18 with the new Quarkus system.

**Reply by darius-m** [reply]
I will preface with the fact that I have not yet tried the Quarkus version, so I am not sure how well the configuration is written (it looks good to me from what I have seen; I am not sure how the `proxy` parameter works, however).

You should at least try to check that HTTPS works on the Keycloak server - try to connect to the service on port `8443` using:
 - `curl` or `wget` to fetch data;
 - `openssl` or `gnutls-cli` (or similar tools) to get information about the certificate / TLS parameters.

I would try to check the connection both from the system Apache connects from (and the Keycloak server, if it was a separate server).

From the error thrown by Apache, I expect there may be an error with the certificate or TLS parameters. From the documentation, I see that there is a parameter to configure the TLS ciphers, which you have not configured yet. Since the TLSv1.2 protocol is also configured (which differs from the default of TLSv1.3 only), there may be an issue with the ciphers.

I am not sure what the expected format is, but some recommended TLS protocol / cipher combinations can be found on the [Mozilla SSL configuration generator](https://ssl-config.mozilla.org/) site. Keycloak may expect the names to be in a format that is compatible with a Java TLS library, so a different naming conventions may need to be used.

**Reply by ajquick** [reply]
@darius-m  Doh! That makes sense. It doesn't have a certificate for 127.0.0.1, it just has one for the hostname keycloak.domain.com.

> curl: (51) SSL: no alternative certificate subject name matches target host name '127.0.0.1'
> depth=0 CN = keycloak.domain.com
> 139815364945728:error:14094412:SSL routines:ssl3_read_bytes:sslv3 alert bad certificate:ssl/record/rec_layer_s3.c:1544:SSL alert number 42

I guess I need to setup a virtual host for ```<VirtualHost 127.0.0.1:8443>``` and feed that a valid certificate for the IP? Or possibly just use the correct certificate in the previous VirtualHost example?

**Reply by darius-m** [reply]
OK, I was honestly expecting a more arcane issue, since invalid certificate domains should be a common enough problem to be reported using a clearer error.

For internal communication I would use a self-signed certiifcate that specifies whatever domain names / IPs the Keycloak server(s) have in the `subjectAltName` extension (e.g., using the `-addext` parameter to `openssl`), and then add have the reverse proxy trust the CA's certificate (or the self-signed certificate itself). I know Nginx and HAProxy support such options, so I would assume Apache does as well.

**Reply by ajquick** [reply]
I've abandoned the idea of using Apache for Keycloak. Cpanel now has support for Nginx running ahead of Apache, so I installed a package called [Engintron](https://github.com/engintron/engintron) that will feed Keycloak and all other dynamic content will go to Apache. I still need to grab the internal certificate, but so far I got Keycloak with Edge proxy running with just a few lines in Engintron.

**Reply by ZPascal** [reply]
Hi @ajquick, since a few days, I've switched to a TLS connection between the Proxy and the KeyCloak server. I've used the following configuration to establish the corresponding connection. BTW: I've skipped the CA validation, because for my special case it's not necessary to use the CA validation. I hope that helps the community to solve the problem.

**KeyCloak configuration:**
```conf
...

# Enable proxy mode
proxy=reencrypt
http-relative-path=/auth

https-certificate-file=/test/keycloak/conf/certs/crt.pem
https-certificate-key-file=/test/keycloak/conf/certs/key.pem
https-port=8443
```

**Apache2 configuration:**
```conf
<IfModule mod_ssl.c>
<VirtualHost *:443>
        ServerName sso.test.com
        ServerAdmin root@localhost
        DocumentRoot /var/www/html

        # don't lose time with IP address lookups
        HostnameLookups Off
        # needed for named virtual hosts
        UseCanonicalName Off

        ProxyPreserveHost On
        RequestHeader set X-Forwarded-Proto "https"
        RequestHeader set X-Forwarded-Port "443"
        ProxyPass / https://localhost:8443/
        ProxyPassReverse / https://localhost:8443/
        
        SSLProxyEngine on
        SSLProxyCACertificateFile /test/keycloak/conf/certs/ca.crt
        SSLProxyCheckPeerCN off

        ProxyRequests Off

        SecRuleEngine On
        SecRule ARGS:triggerparam "@contains trigger" "id:1234,deny,status:403,msg:'Our trigger rule has triggered'"

        <LocationMatch "(/auth/admin/|/auth/realms/master/)$">
                deny from all
                allow from x.x.x.x

                AuthType Basic
                AuthName "SSO Admin protection"
                AuthUserFile /path/sso/.htpasswd
                require valid-user
        </LocationMatch>

        <LocationMatch "/auth/admin">
                SecRuleRemoveById 911100
                SecRuleRemoveById 949110
                SecRuleRemoveById 980130
        </LocationMatch>

        ErrorLog /var/log/apache2/keycloaktestcom.log
        LogLevel warn
        CustomLog /var/log/apache2/keycloaktestcom.log combined
        ServerSignature Off

SSLCertificateFile /etc/letsencrypt/live/sso.test.com/fullchain.pem
SSLCertificateKeyFile /etc/letsencrypt/live/sso.test.com/privkey.pem
Include /etc/letsencrypt/options-ssl-apache.conf
</VirtualHost>
</IfModule>
```

**Comment by pedroigor** [comment]
Hey guys, what do you think we can improve to make it easier? Docs?

**Reply by thegushi** [reply]
You asked, two years ago, "what do you think we can improve to make it easier".

The very first line of the discussion says: "I wasn't able to find a working example".  You have example docs here: https://www.keycloak.org/server/reverseproxy

Please just take that pages, and put a sample vhost config into it, rather than prattling on about a bunch of deprecated config options which *could* be set.  List an example for where Keycloak is running on localhost with no SSL, and list an example where keycloak is doing SSL internally (some orgs require this, I guess?).

That's what I think you can do.

**Comment by wolfgang-werner** [comment]
Hi, I had a similar issue, **solved by setting hostname-strict-https to true** configuration is now working as shown. 

Running keycloak 19.0.3 behind apache 2.4.41 / LetsEncrypt on Ubuntu 20.04.1 (aws Lightsail).

I secured the apache proxy server with correct SSL certificate and would like to terminate SSL on the apache, forwarding to key cloak by plain http (localhost).

Konfiguration keycloak
```
proxy=edge
http-enabled=true
http-host=127.0.0.1
http-port=8888
http-relative-path=/
hostname=keycloak.example.de
hostname-strict-https=true
```

Konfiguration apache
```
<VirtualHost *:443>
    ServerAdmin admin@example.de
    ServerName  keycloak.example.de
    DocumentRoot /var/www/keycloak.example.de
    ErrorLog ${APACHE_LOG_DIR}/keycloak-error.log
    CustomLog ${APACHE_LOG_DIR}/keycloak-access.log combined
    LogLevel Info
    ServerSignature Off
    HostnameLookups Off
    UseCanonicalName Off
    ProxyPreserveHost On
    RequestHeader set "X-Forwarded-Proto" "https"
    RequestHeader set "X-Forwarded-Port" "443"
    RequestHeader set x-ssl-client-cert "%{SSL_CLIENT_CERT}s"
    ProxyPass / http://localhost:8888/
    ProxyPassReverse / http://localhost:8888/
    ProxyRequests Off
    SSLCertificateFile /etc/letsencrypt/live/keycloak.example.de/fullchain.pem
    SSLCertificateKeyFile /etc/letsencrypt/live/keycloak.example.de/privkey.pem
    Include /etc/letsencrypt/options-ssl-apache.conf
</VirtualHost>
```

When I hit the base url, the Welcome screen is shown. When I try to access the Admin Console, I see the following successful requests, so initial loading page and javascript are loaded correctly. However, after this the admin page's loading indicator does not go away. Inside the Page I see the following configuration:

```
<script id="environment" type="application/json">
    {
      "loginRealm": "master",
      "authServerUrl": "https://keycloak.example.de:443",
      "authUrl": "https://keycloak.example.de:443",
      "consoleBaseUrl": "/admin/master/console/",
      "resourceUrl": "/resources/sqe7y/admin/keycloak.v2",
      "masterRealm": "master",
      "resourceVersion": "sqe7y",
      "commitHash": "b87a4c18f6fd788de71e5bfc02c95d2c88763a6d",
      "isRunningAsTheme": true
    }
  </script>
```  

From debugging with Developer Extensions in chrome, I see:

```  
https://keycloak.example.de/realms/master/protocol/openid-connect/3p-cookies/step1.html    --> 200
https://keycloak.example.de/realms/master/protocol/openid-connect/3p-cookies/step2.html    --> 200
https://keycloak.example.de/realms/master/protocol/openid-connect/login-status-iframe.html --> 200
https://keycloak.example.de/realms/master/protocol/openid-connect/login-status-iframe.html/init?client_id=security-admin-console&origin=https%3A%2F%2Fkeycloak.example.de --> 204
```  

Questions:
- What can I do to debug this? 
- What can I do to resolve this?

**Reply by ZPascal** [reply]
Hi @wolfgang-werner, could you please also share the Apache2 logs for the corresponding VHost entry? FYI: I've got a working solution based on SSL between Apach2 proxy and Keycloak (v19.0.3). Is this also a possible solution for you?

**Reply by wolfgang-werner** [reply]
Hi Pascal, thanks for picking this up. My configuration is now working as expected. I somehow missed the configuration parameter hostname-strict-https=true. I like to keep it as it is now as it keeps me from creating internal certificates and is a localhost connection only. 

I had a look into the logfiles and without additional proxy log configuration they showed the mentioned url calls without further detail.

**Reply by ZPascal** [reply]
Hi @wolfgang-werner, good to hear that everything works now as expected.

**Comment by liv-io** [comment]
Thank you @ZPascal and @wolfgang-werner. This was very valuable information.

**Comment by gerdesj** [comment]
I eventually noticed an error in the web tools console about X-Frame-Options.

I added this to my apache config:

`Header set X-Frame-Options "SAMEORIGIN"`

... and a long debugging session comes to an end.  I have Keycloak 23.0.1 running on 127.0.0.1:8080 and Apache on https://systemname.example.com:8443.  Until I added the X-Frame-Options header the  --hostname-debug=true showed all valid settings.

**Comment by gerdesj** [comment]
Putting it all together:

Apache Vhost:
```
Listen 10.77.10.10:8443

<VirtualHost 10.77.10.10:8443>
        ServerName   auth.example.co.uk
        ServerAdmin  admin@example.co.uk
        DocumentRoot /var/www/html

        ErrorLog  /var/log/apache2/keycloak-error.log
        LogLevel  warn
        CustomLog /var/log/apache2/keycloak.log combined

        SSLEngine on
        SSLCertificateFile      /etc/ssl/www.pem
        SSLCertificateKeyFile   /etc/ssl/www.key
        SSLCertificateChainFile /etc/ssl/ca.pem

        ServerSignature   Off
        HostnameLookups   Off
        ProxyPreserveHost On

        Header unset X-Frame-Options

        ProxyPass     / http://localhost:8080/
        ProxyRequests Off

</VirtualHost>
```

Partial keycloak.conf:

```
proxy=edge
http-host=127.0.0.1
http-enabled=true
http-port=8080

hostname-url=https://auth.example.co.uk:8443
hostname-admin-url=https://auth.example.co.uk:8443

```

systemctl edit --full keycloak.conf
```

[Unit]
Description=Keycloak Application Server
After=syslog.target network.target

[Service]
Type=idle
User=keycloak
Group=keycloak
RemainAfterExit=yes

# Create /opt/keycloak/data and chown keycloak:keycloak
Environment="KC_TRANSACTION_XA_ENABLED=true"
Environment="QUARKUS_TRANSACTION_MANAGER_ENABLE_RECOVERY=true"
ExecStart=/opt/keycloak/bin/kc.sh start --optimized 

[Install]
WantedBy=multi-user.target
```

(Edited.  Something was setting the X-Frame-Options header, hence now using the unset.)
